arbnode/transaction_streamer.go

```diff
@@ -675,14 +675,22 @@ func endBatch(batch ethdb.Batch) error {
 	return batch.Write()
 }
 
-func (s *TransactionStreamer) AddMessagesAndEndBatch(pos arbutil.MessageIndex, messagesAreConfirmed bool, messages []arbostypes.MessageWithMetadata, batch ethdb.Batch) error {
+func (s *TransactionStreamer) AddMessagesAndEndBatch(pos arbutil.MessageIndex, messagesAreConfirmed bool, messages []arbostypes.MessageWithMetadata, blockMetadataArr []arbostypes.BlockMetadata, batch ethdb.Batch) error {
 	messagesWithBlockInfo := make([]arbostypes.MessageWithMetadataAndBlockInfo, 0, len(messages))
 	for _, message := range messages {
 		messagesWithBlockInfo = append(messagesWithBlockInfo, arbostypes.MessageWithMetadataAndBlockInfo{
 			MessageWithMeta: message,
 		})
 	}
 
+	if len(blockMetadataArr) == len(messagesWithBlockInfo) {
+		for i, blockMetadata := range blockMetadataArr {
+			messagesWithBlockInfo[i].BlockMetadata = blockMetadata
+		}
+	} else if len(blockMetadataArr) > 0 {
+		log.Warn("Size of blockMetadata array doesn't match the size of messages array", "lockMetadataArrSize", len(blockMetadataArr), "messagesSize", len(messages))
+	}
```

Comment:
I think this should be an error return here. Even if there's no metadata, blockMetadataArr should be an array of nils the same size as messagesWithBlockInfo.

Reply:
This isn't the case when InboxTracker adds messages to the batch https://github.com/OffchainLabs/nitro/blob/9a85c95f834764ffa04a6da3275e077f3f36ee38/arbnode/inbox_tracker.go#L826-L830
in which case len(blockMetadataArr) is zero but we should accept that as a correct input. Similarly a bunch of cases with invoking of `AddMessages` function of txStreamer.

But I agree that we should return an error here instead of just logging, because in normal conditions wrt blockMetadataArr (when called from seq-coordinator) what you said is absolutely true